src/playlist-maker.js

```diff
@@ -248,21 +245,27 @@ export default function factory(player, initialList, initialIndex = 0) {
 
       // @todo - Simplify this to `list.slice()` for v5.
       const previousPlaylist = Array.isArray(list) ? list.slice() : null;
+      const nextPlaylist = newList.slice();
 
-      list = newList.slice();
+      list = nextPlaylist.slice();
 
       // Transform any primitive and null values in an input list to objects
       if (list.filter(item => isItemObject(item)).length !== list.length) {
         list = transformPrimitiveItems(list);
       }
 
+      // Add unique id to each playlist item. This id will be used
+      // to determine index in cases where there are more than one
+      // identical sources in the playlist.
+      generatePlaylistItemId(list);
```

Comment:
This was moved up because these IDs need to be populated _before_ we call `currentItem()` below for the `newIndex`.